File: packages/flags-java/src/main/java/com/signakit/flags/audience/AudienceMatcher.java
package com.signakit.flags.audience;

import com.signakit.flags.config.ConfigModels.AudienceCondition;
import com.signakit.flags.config.ConfigModels.AudienceMatchType;
import com.signakit.flags.config.ConfigModels.ConfigRuleAudience;

import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Operator + audience match-type evaluation. Mirrors
 * {@code packages/flags-node/src/audience-matcher.ts}.
 */
public final class AudienceMatcher {
    private AudienceMatcher() {}

    public static boolean matchesCondition(AudienceCondition condition, Map<String, Object> attributes) {
        if (attributes == null) return false;
        Object userValue = attributes.get(condition.attribute);
        if (userValue == null) return false;

        String operator = condition.operator;
        Object value = condition.value;

        return switch (operator) {
            case "equals" -> Objects.equals(userValue, value);
            case "not_equals" -> !Objects.equals(userValue, value);
            case "greater_than" -> compareNumbers(userValue, value, c -> c > 0);
            case "less_than" -> compareNumbers(userValue, value, c -> c < 0);
            case "greater_than_or_equals" -> compareNumbers(userValue, value, c -> c >= 0);
            case "less_than_or_equals" -> compareNumbers(userValue, value, c -> c <= 0);
            case "in" -> value instanceof List<?> list && list.contains(userValue);
            case "not_in" -> !(value instanceof List<?> list) || !list.contains(userValue);
            case "contains" -> containsOp(userValue, value);
            case "not_contains" -> notContainsOp(userValue, value);
            default -> {
                System.err.println("[SignaKit] Unknown operator: " + operator);
                yield false;
            }
        };
    }

    private interface IntPred {
        boolean test(int v);
    }

    private static boolean compareNumbers(Object a, Object b, IntPred pred) {
        if (a instanceof Number na && b instanceof Number nb) {
            int cmp = Double.compare(na.doubleValue(), nb.doubleValue());
            return pred.test(cmp);
        }
        return false;
    }

    private static boolean containsOp(Object userValue, Object value) {
        if (userValue instanceof String s && value instanceof String v) {
            return s.contains(v);
        }
        if (userValue instanceof List<?> list && value instanceof String v) {
            return list.contains(v);
        }
        return false;
    }

    private static boolean notContainsOp(Object userValue, Object value) {
        if (userValue instanceof String s && value instanceof String v) {
            return !s.contains(v);
        }
        if (userValue instanceof List<?> list && value instanceof String v) {
            return !list.contains(v);
        }
        return true;
    }

    public static boolean matchesAudience(ConfigRuleAudience audience, Map<String, Object> attributes) {
        if (audience.conditions == null || audience.conditions.isEmpty()) return true;
        for (AudienceCondition c : audience.conditions) {
            if (!matchesCondition(c, attributes)) return false;
        }
        return true;
    }

    public static boolean matchesAudiences(
            List<ConfigRuleAudience> audiences,
            AudienceMatchType matchType,
            Map<String, Object> attributes) {
        if (audiences == null || audiences.isEmpty()) return true;
        if (matchType == AudienceMatchType.ANY) {
            for (ConfigRuleAudience a : audiences) {
                if (matchesAudience(a, attributes)) return true;
            }
            return false;
        }
        // Default ALL semantics
        for (ConfigRuleAudience a : audiences) {
            if (!matchesAudience(a, attributes)) return false;
        }
        return true;
    }
}

File: packages/flags-php/src/AudienceMatcher.php
<?php

declare(strict_types=1);

namespace SignaKit\FlagsPhp;

/**
 * Evaluates audience conditions against a user's attributes.
 */
final class AudienceMatcher
{
    /**
     * @param array<int, array<string, mixed>>  $audiences       Audience definitions from the rule
     * @param string                            $matchType       'any' (OR) or 'all' (AND)
     * @param array<string, mixed>              $userAttributes  Caller-supplied user attributes
     */
    public static function matchesAudiences(
        array $audiences,
        string $matchType,
        array $userAttributes,
    ): bool {
        // No audiences defined → matches all users
        if (count($audiences) === 0) {
            return true;
        }

        foreach ($audiences as $audience) {
            /** @var array<int, array<string, mixed>> $conditions */
            $conditions     = (array) ($audience['conditions'] ?? []);
            $audienceMatch  = self::evaluateConditions($conditions, $userAttributes);

            if ($matchType === 'any' && $audienceMatch) {
                return true;
            }

            if ($matchType === 'all' && !$audienceMatch) {
                return false;
            }
        }

        // 'any' with no match → false; 'all' with all match → true
        return $matchType === 'all';
    }

    /**
     * All conditions in a single audience must pass (AND within an audience).
     *
     * @param array<int, array<string, mixed>> $conditions
     * @param array<string, mixed>             $userAttributes
     */
    private static function evaluateConditions(array $conditions, array $userAttributes): bool
    {
        foreach ($conditions as $condition) {
            if (!self::matchesCondition($condition, $userAttributes)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Evaluate a single condition against the user's attributes.
     *
     * @param array<string, mixed> $condition
     * @param array<string, mixed> $userAttributes
     */
    public static function matchesCondition(array $condition, array $userAttributes): bool
    {
        $attribute = (string) ($condition['attribute'] ?? '');
        $operator  = (string) ($condition['operator']  ?? '');
        $expected  = $condition['value'] ?? null;

        if (!array_key_exists($attribute, $userAttributes)) {
            return false;
        }

        $actual = $userAttributes[$attribute];

        return match ($operator) {
            'equals'                  => $actual == $expected,
            'not_equals'              => $actual != $expected,
            'greater_than'            => is_numeric($actual) && is_numeric($expected) && (float) $actual >  (float) $expected,
            'less_than'               => is_numeric($actual) && is_numeric($expected) && (float) $actual <  (float) $expected,
            'greater_than_or_equals'  => is_numeric($actual) && is_numeric($expected) && (float) $actual >= (float) $expected,
            'less_than_or_equals'     => is_numeric($actual) && is_numeric($expected) && (float) $actual <= (float) $expected,
            'in'                      => is_array($expected) && in_array($actual, $expected, strict: false),
            'not_in'                  => is_array($expected) && !in_array($actual, $expected, strict: false),
            'contains'                => is_string($actual) && is_string($expected) && str_contains($actual, $expected),
            'not_contains'            => is_string($actual) && is_string($expected) && !str_contains($actual, $expected),
            default                   => false,
        };
    }
}
